lens/sqlrepo/blockstore.go

```diff
@@ -0,0 +1,235 @@
+package sqlrepo
+
+import (
+	"context"
+	"strings"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/blockstore"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multibase"
+)
+
+var errNoRows = pgx.ErrNoRows
+
+var dbpool *pgxpool.Pool
+
+var log = logging.Logger("sql")
+
+func DB(connStr string) *pgxpool.Pool {
+	if dbpool == nil {
+		var err error
+		dbpool, err = pgxpool.Connect(context.Background(), connStr)
+		if err != nil {
+			log.Fatalf("failed to connect to %s: %s", connStr, err)
+		}
+
+		for _, ddl := range []string{
+			"CREATE TABLE IF NOT EXISTS blocks(" +
+				"multiHash TEXT NOT NULL PRIMARY KEY," +
+				"initialCodecID INTEGER NOT NULL," +
+				"content BYTEA NOT NULL" +
+				")",
+			"CREATE TABLE IF NOT EXISTS heads(" +
+				"seq SERIAL NOT NULL PRIMARY KEY," +
+				"ts TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL ," +
+				"height BIGINT NOT NULL," +
+				"blockCids TEXT NOT NULL" +
+				")",
+			"CREATE INDEX IF NOT EXISTS height_idx ON heads ( height )",
+		} {
+			if _, err = dbpool.Exec(context.Background(), ddl); err != nil {
+				log.Fatalf("On-connect DDL execution failed: %s", err)
+			}
+		}
+	}
+
+	return dbpool
+}
+
+type SqlBlockstore struct {
+	db *pgxpool.Pool
+}
+
+func NewBlockStore(connstr string) (blockstore.Blockstore, error) {
+
+	sbs := &SqlBlockstore{
+		db: DB(connstr),
+	}
+
+	// we do not currently use the Identity codec, but just in case...
+	return blockstore.WrapIDStore(sbs), nil
+}
+
+func keyFromCid(c cid.Cid) (k string) {
+	// CB: we can not store complete CIDs as keys - code expects being able
+	// to match on multihash alone, likely both over raw *and* CBOR :( :( :(
+	//return c.Encode(multibase.MustNewEncoder(multibase.Base64))
+
+	k, _ = multibase.Encode(multibase.Base64, c.Hash())
+	return
+}
+
+func (sbs *SqlBlockstore) Has(c cid.Cid) (has bool, err error) {
+	err = sbs.db.QueryRow(
+		context.Background(),
+		"SELECT EXISTS( SELECT 42 FROM blocks WHERE multiHash = $1 )",
```

Comment:
42?

Reply:
i think this is just for the where, and it's just a constant that we check if any rows are returned.

Reply:
Could I bother you to add a comment mentioning that?

Reply:
We can change it to 23 if you guys *really* insist...